libqaul/common/src/payload.rs

```diff
@@ -0,0 +1,89 @@
+//! Types related to data being moved around and its verification.
+use crate::HashBytes;
+use blake2::VarBlake2b;
+use blake2::digest::{VariableOutput, Input};
+use subtle::Choice;
+use generic_array::GenericArray;
+use std::convert::AsMut;
+
+fn copy_into_array<A, T>(slice: &[T]) -> A where
+    A: Default + AsMut<[T]>,
+    T: Copy,
+{
+    let mut a = Default::default();
+    <A as AsMut<[T]>>::as_mut(&mut a).copy_from_slice(slice);
+    a
+}
+
+/// The actual content of a message, along with the mechanism to validate that
+/// no transmission errors occurred (message digest).
+#[derive(PartialEq, Eq, Debug)]
+pub struct Payload {
+    // TODO DESIGN: add a payload type field?
+    // Upcast to u64 so that there is no issue knowing what the length of a Payload is when sent to different platforms,
```

Comment:
We could totally do an ASN.1 and have variable length lengths :joy: (please ignore me)

Using a 64bit unsigned for length sounds like a good idea. Payload type could probably just be another string, similar to the service. Not something _we_ as library developers really care about but something where service authors can do stuff with. Good idea :+1:

Reply:
After looking into this I think maybe the better place for this functionality is to have a "message type" field in the Message.